sql-plugin/src/main/scala/com/nvidia/spark/rapids/TypeChecks.scala

```diff
@@ -801,7 +801,9 @@ object CaseWhenCheck extends ExprChecks {
  * This is specific to WidowSpec, because it does not follow the typical parameter convention.
  */
 object WindowSpecCheck extends ExprChecks {
-  val check: TypeSig = TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL
+  val check: TypeSig =
+    TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL +
+      TypeSig.STRUCT.nested(TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL)
```

Comment:
So we only support single level structs for lead/lag?

Reply:
Not at all, sir.

`Struct<Struct>`:
```scala
scala> val window = " partition by uid order by dt ";
scala> sql(s" select  lead(struct(struct(dollars, dateStr)), 3) over ($window) lead_3 from mytable ").explain
GpuColumnarToRow false
+- GpuProject [lead_3#178]
   +- GpuWindow [_w0#179, uid#144L, dt#147, gpulead(_w0#179, 3, null) gpuwindowspecdefinition(uid#144L, dt#147 ASC NULLS FIRST, gpuspecifiedwindowframe(RowFrame, 3, 3)) AS lead_3#178], [uid#144L], [dt#147 ASC NULLS FIRST]
      +- GpuCoalesceBatches batchedbykey(uid#144L ASC NULLS FIRST)
         +- GpuSort [uid#144L ASC NULLS FIRST, dt#147 ASC NULLS FIRST], false, com.nvidia.spark.rapids.OutOfCoreSort$@7342eb53
            +- GpuShuffleCoalesce 2147483647
               +- GpuColumnarExchange gpuhashpartitioning(uid#144L, 200), ENSURE_REQUIREMENTS, [id=#537]
                  +- GpuProject [named_struct(col1, named_struct(dollars, dollars#149, dateStr, dateStr#148)) AS _w0#179, uid#144L, dt#147]
                     +- GpuFileGpuScan orc [uid#144L,dt#147,dateStr#148,dollars#149] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex[file:/home/mithunr/workspace/Demos/windowbench/data/gs/1/part-00000.snappy.orc], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<uid:bigint,dt:date,dateStr:string,dollars:int>
```

`Struct<Array>`:
```scala
scala> sql(s" select  lead(struct(array(dollars)), 3) over ($window) lead_3 from mytable ").explain
== Physical Plan ==
GpuColumnarToRow false
+- GpuProject [lead_3#191]
   +- GpuWindow [_w0#192, uid#144L, dt#147, gpulead(_w0#192, 3, null) gpuwindowspecdefinition(uid#144L, dt#147 ASC NULLS FIRST, gpuspecifiedwindowframe(RowFrame, 3, 3)) AS lead_3#191], [uid#144L], [dt#147 ASC NULLS FIRST]
      +- GpuCoalesceBatches batchedbykey(uid#144L ASC NULLS FIRST)
         +- GpuSort [uid#144L ASC NULLS FIRST, dt#147 ASC NULLS FIRST], false, com.nvidia.spark.rapids.OutOfCoreSort$@7342eb53
            +- GpuShuffleCoalesce 2147483647
               +- GpuColumnarExchange gpuhashpartitioning(uid#144L, 200), ENSURE_REQUIREMENTS, [id=#632]
                  +- GpuProject [named_struct(col1, array(dollars#149)) AS _w0#192, uid#144L, dt#147]
                     +- GpuFileGpuScan orc [uid#144L,dt#147,dollars#149] Batched: true, DataFilters: [], Format: ORC, Location: InMemoryFileIndex[file:/home/mithunr/workspace/Demos/windowbench/data/gs/1/part-00000.snappy.orc], PartitionFilters: [], PushedFilters: [], ReadSchema: struct<uid:bigint,dt:date,dollars:int>
```

The issue is the same as mentioned [above](https://github.com/NVIDIA/spark-rapids/pull/2842#issuecomment-871753188): The sorting of the outputs for the result comparison causes the test to fail spuriously, because `Sort` multi-level structs isn't on GPU. :/

Reply:
`@ignore_order(local=True)` does the sorting in python on the CPU. It works with multi-level structs, arrays, and I think I made it works with maps too.

Reply:
So if we can test it will that let you turn it on?

Reply:
Yes, sir. I'll add back the test and check with `local=true`. Thanks for the tip.

Reply:
Just confirmed that `@ignore_order(local=True)` works. 

Added a separate test for `struct<array>`, `array<struct<array>`, `struct<struct<array>>`, etc. I didn't fold this back into `test_multi_types_window_aggs_for_rows_lead_lag` because of the `ORDER BY c` clause, which will punt the operator off the GPU.